core/server/master/src/main/java/alluxio/master/meta/JournalSpaceMonitor.java

```diff
@@ -0,0 +1,176 @@
+ * This monitor is only applicable to systems which are running on the Linux OS and which are
+ * running the embedded journal. Other configurations are invalid. Running this monitor on such
+ * systems is undefined behavior.
+ */
+public class JournalSpaceMonitor implements HeartbeatExecutor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JournalSpaceMonitor.class);
+
+  /**
+   * Convenience method for {@link #JournalSpaceMonitor(String, long)}.
+   *
+   * @param configuration the current alluxio configuration
+   */
+  public JournalSpaceMonitor(AlluxioConfiguration configuration) {
+    this(configuration.get(PropertyKey.MASTER_JOURNAL_FOLDER), configuration.getLong(
+        PropertyKey.MASTER_JOURNAL_SPACE_MONITOR_PERCENT_FREE_THRESHOLD));
+  }
+
+  private final String mJournalPath;
+  private final long mWarnCapacityPercentThreshold;
+
+  // Used to store information for metrics gauges
+  private AtomicReference<Map<String, JournalDiskInfo>> mMetricInfo = new AtomicReference();
+
+  /**
+   *
+   * @param journalPath the path to the journal location
+   * @param warnCapacityPercentThreshold when a device utilization falls below this threshold,
+   *                                     emit warnings to the log
+   */
+  public JournalSpaceMonitor(String journalPath, long warnCapacityPercentThreshold) {
+    Preconditions.checkArgument(Files.exists(Paths.get(journalPath)),
+        "journal path must exist");
+    mJournalPath = journalPath;
+    mWarnCapacityPercentThreshold = warnCapacityPercentThreshold;
+  }
+
+  /**
+   * Gets the raw output from the {@code df} command.
+   *
+   * This method is used to allow testing of the JournalSpaceMonitor code.
+   *
+   * @return the output from the {@code df} command
+   * @throws IOException if retrieving the output fails
+   */
+  @VisibleForTesting
+  CommandReturn getRawDiskInfo() throws IOException {
+    return ShellUtils.execCommandWithOutput("df", "-k", "-P", mJournalPath);
+  }
+
+  /**
+   * Gets the current disk information from the system for the journal path.
+   *
+   * @return a list of the disks that back the journal path. Many systems will just have one
+   * disk. Depending on the configuration multiple disks may back the journal path, in which
+   * case we report information on all of them
+   * @throws IOException when obtaining the disk information fails
+   */
+  protected synchronized List<JournalDiskInfo> getDiskInfo() throws IOException {
+    // command allowed by the POSIX specification
+    CommandReturn output = getRawDiskInfo();
+    if (output.getExitCode() != 0) {
+      throw new IOException("Command failed with exit code " + output.getExitCode());
+    }
+    List<JournalDiskInfo> infos = Arrays.stream(output.getOutput().split("\n"))
+        .skip(1)
+        .map(infoLine -> Arrays.stream(infoLine.split(" "))
+              .map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList()))
+        .filter(data -> data.size() >= 6)
```

Comment:
What does the output look like?

Also, I'm worried about parsing `df` output, because it could be fragile?

Reply:
df output is not fragile. The output is actually defined by a POSIX standard: https://pubs.opengroup.org/onlinepubs/000095399/utilities/df.html

Additionally, it's part of coreutils which is generally unchanging. I think that generally, this should be safe to do. I looked into the coreutils source code for `df` and found that there's a lot of corner cases that the implementation accounts for that I would unlikely be able to do. Since `df` is included in essentially all distributions and is defined by the POSIX standard my conclusion was parsing the output should be safe.